File: OpenMP/dependency.c
#include <stdio.h>

int main () {

    int N = 12000; 
    /*
      Declare a series of arrays.
      AS is for serial version, AP for the parallel method.
    */
    double AS[N][N], AP[N][N];
    double X = 3;
    /*
      Variable set if error found with parallel version
    */
    int error = 0;

    int i, j, k;

    /*
      ARRAY INITIALISATION
    */

    for(j=0;j<N;j++) {
	for (i=0;i<N;i++) {
	    AP[i][j] = i * j;
	    AS[i][j] = i * j;
	}
    }

    /*
      PARALLELISE THIS SECTION
      Serial version to be parallelised
    */
    for(k=0;k<5;k++) {
	for(i=1;i<N;i++) {
	    for(j=0;j<N;j++) {
		AP[i][j] = AP[i-1][j] + X + (double)k;
	    }
	}
    }
    /*
      SERIAL COMPUTATION
      Serial version for testing correctness
    */
    for(k=0;k<5;k++) {
	for(i=1;i<N;i++) {
	    for(j=0;j<N;j++) {
		AS[i][j] = AS[i-1][j] + X + (double)k;
	    }
	}
    }


    /*
      TEST FOR SUCCESS 
      Compare results of each parallel version with the serial results.
    */
    
    for(j=0;j<N;j++) {
	for(i=1;i<N;i++) {
	    if(AP[i][j] != AS[i][j]) {
		error = 1;
	    }
	}
    }

    if (error == 0)
	printf("No error encountered with the parallelisation.\n");
    else
	printf("Error encountered with the parallelisation.\n");
    

    return 0;

}

File: OpenMP/dependency.f90
PROGRAM Data_Depend
  IMPLICIT NONE
  INCLUDE "omp_lib.h"
  INTEGER, PARAMETER :: N = 12000
  REAL, DIMENSION(N,N) :: AP, AS
  REAL :: x
  INTEGER :: i,j,k, check_error

! intialize the matrices
! AS is for serial version, AP for the parallel method.
  CALL random_number(AP)
  AS=AP
! and the constant
  CALL random_number(x)

! PARALLISE THIS SECTION

  DO k = 1, 5
     DO j = 2, N
        DO  i = 1, N
           AP(i,j) = AP(i,j-1) + x + REAL(k)
        END DO
     END DO
  END DO

! Serial version to compare

  DO k = 1, 5
     DO j = 2, N 
        DO i = 1, N 
           AS(i,j) = AS(i,j-1) + x + REAL(k)
        END DO
     END DO
  END DO

    
! TEST FOR SUCCESS 
! Compare results of parallel version with the serial results.
 
  check_error = 0
  DO i=1,N 
     DO j=1,N
        IF( AP(i,j).ne.AS(i,j) ) THEN
           check_error = 1
        END IF
    END DO
 END DO


 IF (check_error == 0) THEN
    WRITE(*,*) 'No error encountered with solution'
 ELSE
    WRITE(*,*) 'Error encountered with solution'
 END IF
    

END PROGRAM Data_Depend
